### DISCUSSION TITLE
Tuple Parameter Name Enforcement Warning - Missing

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/roslyn#59810] [General]
For the case of Tuple-usage, I believe the compiler SHOULD issue a warning for the following code:
![image](https://user-images.githubusercontent.com/10914515/156018211-5a13de02-caea-4e7e-926a-9dad8c87ee5f.png)

There is strong argument for treating Tuple Field Names the same as Class-Property-Names where mismatches in names can produce a compiler Error  (or Warning at minimum).

Use of Tuples should not exempt you from using the exact Field Names, the same as if you were using a named Class.

(EDITED: To make a good code example, using EventHandler<T> for an event.)

**Comment by CyrusNajmabadi** [comment]
Moving to roslyn as a request for the compiler to detect and warn about a tuple mismatch.

**Comment by CyrusNajmabadi** [comment]
@najak3d I would strongly recommend the use of records. Trying to shoehorn everything into tuples is definitely going to put you in unpleasant situations. 

For tuples things like names are simply a loser suggestion between caller and callee. Both can be different depending on whatever names each side thinks is preferable. When you have multiple values of the same type, you're definitely approaching confusing areas as the type system is no longer protecting you. 

If you want a strong abi that both sides must agree on, then just make such a thing. Use a record struct and you cannot have this issue.

**Comment by najak3d** [comment]
What are the reasons that a Tuple shouldn't be treated the same as an anonymous record?   Which means that use of Tuples would enforce the use of parameter names, unless you opt for the Item1/2/3 generic interface?

For what would be single-use record types, it's nice to avoid "adding a new type to the namespace" and just make it anonymous, defined right where it's used.

```c#
///<summary>Event occurs when Manager changes</summary>
/// <param name="oldMgr">Name of old manager.</param>
/// <param name="newMgr">Name of the new manager.</param>
public event EventHandler<(string oldMgr, string newMgr)> ManagerChanged;
```

Then you avoid the creation of a new strong-typed record, by using this anonymous-record (i.e. Tuple).

In this case, 
1. it's all self-contained, 
2. makes Intellisense more helpful (no F12 to see Def's), 
3. avoids new type creation for one-use, and 
4. Enforces the Field naming, same as if it were a record. 

**I don't see reasons to treat Tuples different from an anonymous Record.   What am I missing?**

**Reply by najak3d** [reply]
In short, how is "use of a one-use record type" superior to use of an "anonymous record"?   (other than the fact that our compiler currently does not currently enforce same-field naming)

**Reply by CyrusNajmabadi** [reply]
> What are the reasons that a Tuple shouldn't be treated the same as an anonymous record?

Because we felt that names there should just be informative, not a part of the binary interface.  We already made those decisions in the past and can certainly not go back on them as that could break lots of code.

> makes Intellisense more helpful

This is not a use case.  You are presuming a specific way that intellisense works.  But intellisense can work however it wants without concern about the language.

> Enforces the Field naming, same as if it were a record.

As mentioned above, this was an explicit decision not to do.

**Reply by najak3d** [reply]
Do you think there is much chance of this mismatch in parameter names triggering a compiler warning? (which could optionally be upgraded to error by dev teams)

I really dislike creating one-use PUBLIC types.  As is, creating these one-use public types is the only way to ensure this type of usage safety.

At best case, you have to create the PUBLIC type right inside the same class as you are using it -- at which point it now pollutes your class's static API with a useless type that no one ever wants to see or use, outside of the context of the event using it.   So now we have TWO API members instead of just one, for everyone to see.

I wish we could rewind time, and make Tuples more strict.  Not sure why they decided leniency in that case was the best choice.

Given the current situation, that we can't undo the past, I think the only viable path to improvement here is to hope this can become a Compiler Error/Warning, which can be disabled on a file-by-file, or project-by-project basis (similar to how the nullable feature was implemented).

**Comment by CyrusNajmabadi** [comment]
> Do you think there is much chance of this mismatch in parameter names triggering a compiler warning? (which could optionally be upgraded to error by dev teams)

I feel like writing an analyzer would work great for this purpose.

**Reply by CyrusNajmabadi** [reply]
> I wish we could rewind time, and make Tuples more strict. Not sure why they decided leniency in that case was the best choice.

Because it's acts as a loose collection of values (similar to parameter lists). As I mentioned above, that's often a much looser contact between caller and callee (and in some places has no enforcement at all).
